go/chat/boxer.go

```diff
@@ -622,7 +626,77 @@ func (b *Boxer) unboxV1(ctx context.Context, boxed chat1.MessageBoxed,
 	}, nil
 }
 
-func (b *Boxer) unboxV2orV3(ctx context.Context, boxed chat1.MessageBoxed,
+func (b *Boxer) validatePairwiseMAC(ctx context.Context, boxed chat1.MessageBoxed) (senderKey []byte, err error) {
+	// First, find a MAC that matches our receiving device encryption KID.
+	ourDeviceKey, err := b.G().ActiveDevice.KeyByType(libkb.DeviceEncryptionKeyType)
+	if err != nil {
+		return nil, err
+	}
+	ourDeviceKeyNacl, ok := ourDeviceKey.(libkb.NaclDHKeyPair)
+	if !ok {
+		return nil, fmt.Errorf("wrong key type, expected NaclDHKeyPair, got %T", ourDeviceKey)
+	}
+	messageMAC, found := boxed.PairwiseMacs[ourDeviceKey.GetKID()]
+	if !found {
+		// This is an error users will actually see when they've just joined a
+		// team or added a new device.
+		return nil, fmt.Errorf("message is not authenticated for this device")
+	}
+
+	// Second, load the device encryption KID for the sender. (We have to go
+	// through the signing KID, because encryption KIDs in the UPAK layout
+	// don't have an associated device.)
+	senderUid, err := keybase1.UIDFromString(boxed.ClientHeader.Sender.String())
+	if err != nil {
+		return nil, err
+	}
+	senderDeviceID, err := keybase1.DeviceIDFromString(boxed.ClientHeader.SenderDevice.String())
+	if err != nil {
+		return nil, err
+	}
+	senderUPAK, _, err := b.G().GetUPAKLoader().LoadV2(libkb.NewLoadUserByUIDArg(ctx, b.G().GlobalContext, senderUid))
+	if err != nil {
+		return nil, err
+	}
+	var senderDeviceSigningKID keybase1.KID
+	for _, key := range senderUPAK.Current.DeviceKeys {
+		if key.DeviceID.Eq(senderDeviceID) && key.Base.IsSibkey {
+			senderDeviceSigningKID = key.Base.Kid
+			break
+		}
+	}
+	var senderDeviceEncryptionKID keybase1.KID
+	for _, key := range senderUPAK.Current.DeviceKeys {
+		if !key.Base.IsSibkey && key.Parent != nil && key.Parent.Equal(senderDeviceSigningKID) {
+			senderDeviceEncryptionKID = key.Base.Kid
+			break
+		}
+	}
+	if senderDeviceEncryptionKID.IsNil() {
+		return nil, fmt.Errorf("failed to find device %s", senderDeviceID.String())
+	}
+	senderDeviceDHKey, err := libkb.ImportKeypairFromKID(senderDeviceEncryptionKID)
+	if err != nil {
+		return nil, err
+	}
+	senderDeviceDHKeyNacl, ok := senderDeviceDHKey.(libkb.NaclDHKeyPair)
+	if !ok {
+		return nil, fmt.Errorf("wrong key type, expected NaclDHKeyPair, got %T", ourDeviceKey)
+	}
+
+	// Finally, validate the MAC.
+	computedMAC, err := makeOnePairwiseMAC(*ourDeviceKeyNacl.Private, senderDeviceDHKeyNacl.Public, boxed.HeaderCiphertext.CanonicalHash())
+	if err != nil {
+		return nil, err
+	}
+	if subtle.ConstantTimeCompare(messageMAC, computedMAC) != 1 {
+		return nil, fmt.Errorf("invalid pairwise MAC")
```

Comment:
should this be a special error as well? the gui may want to display the errors differently?